### DISCUSSION TITLE
Generate text  with stroke

### DISCUSSION THREAD

**Discussion by rebornfz** [libvips/libvips#2123] [Q&A] [answered]
Hi
I need to generate some text with stroke, like this
![t](https://user-images.githubusercontent.com/27042311/109460070-06f26300-7a9b-11eb-8528-670969f0a98b.png)
and write the text to my image.

I want to achieve this use simple c api. I have used vips_text() to generate colored text and simple shadows, like https://github.com/libvips/pyvips/issues/204
How can i add stroke to the text？

**Comment by jcupitt** [comment, marked as answer]
Probably the best solution would be to convert your text to a set of paths and set a stroke in the SVG. You'd need an SVG library to do that, but you could render the SVG to a bitmap with libvips.

You could render to a bitmap and dilate, but I don't know if the quality would be good enough for you. For example:

```python
#!/usr/bin/python3

import sys
import pyvips

text = pyvips.Image.text(sys.argv[2], dpi=600)

radius = 5

# we need to make the text image large enough to hold the extra edges
text = text.embed(radius, radius, 
    text.width + 2 * radius, text.height + 2 * radius)

# dilate the text mask, then apply a slight blur to alias the
# edges
circle_mask = pyvips.Image.black(radius * 2 + 1, radius * 2 + 1) \
    .add(128) \
    .draw_circle(255, radius, radius, radius, fill=True)
wide_text = text.dilate(circle_mask).gaussblur(0.5, min_ampl=0.1)

# paint wide text blue
background_text = wide_text \
    .new_from_image([128, 128, 255]) \
    .bandjoin(wide_text) \
    .copy(interpretation="srgb")

# paint original text black
foreground_text = text \
    .new_from_image([0, 0, 0]) \
    .bandjoin(text) \
    .copy(interpretation="srgb")

# composite black over blue
image = background_text.composite(foreground_text, "over")

image.write_to_file(sys.argv[1])
```

Then run:

```
./outline.py x.png "hello world"
```

To make:

![x](https://user-images.githubusercontent.com/580843/109488062-f92de500-7a7c-11eb-9d66-4fc2c5099403.png)

**Reply by rebornfz** [reply]
thank you very much. 
I want to use c to achieve this. What is dilat() called in c api?

**Reply by jcupitt** [reply]
It's a shortcut for vips_morph() with DILATE set:

https://github.com/libvips/pyvips/blob/master/pyvips/vimage.py#L1282-L1284

https://libvips.github.io/libvips/API/current/libvips-morphology.html#vips-morph

**Reply by rebornfz** [reply]
thank you. I have done like this
![stroke1](https://user-images.githubusercontent.com/27042311/109622672-55752f80-7b77-11eb-9e51-d8037aa2fac3.png)

**Reply by jcupitt** [reply]
Huh, pretty! You can use embed to add more space around the text for the effects and the shadow, I expect you know.

**Comment by rebornfz** [comment]
@jcupitt 
Hello,
I found the stroke is rough and jagged after zooming in. 
How can I make the stroke look better？

**Reply by jcupitt** [reply]
The edges won't be anti-aliased, if that's what you mean.

You could render at 2x resolution and then shrink, would that work?

Alternatively, you can render the stroke with SVG, though that would need an SVG editor.

**Reply by rebornfz** [reply]
@jcupitt  How can I use svg to render the stroke? Now I'm using the Pangomark to render text.

**Reply by jcupitt** [reply]
You'd need to do it in an SVG editor like inkscape. Turn the text into a set of paths, remove the fill and set the stroke width to be a large value. Finally render the original text on top again, and save as SVG.

If that's not possible, you can try rendering at 2x magnification, dilating, and then shrinking. It ought to look OK, though probably not quite as good.

**Reply by rebornfz** [reply]
Which function is faster to zoom in and out?

**Reply by jcupitt** [reply]
Sorry, I don't understand. Zoom in and out with what?

**Reply by rebornfz** [reply]
Rendering at 2x magnification and shrinking. My poor english

**Reply by jcupitt** [reply]
Ah no problem. You'd need to benchmark it and also check quality.

**Comment by machineonamission** [comment]
I found that the dilate effect's aliasing was unpleasant and didn't go away with scaling. Below is code I wrote using the gaussian blur effect + basic multiplication to achieve a much better effect
```py
def outline_gaussthreshold(image: pyvips.Image, radius: int | None = None,
                           color: typing.Sequence[int] | None = None) -> pyvips.Image:
    # blur the alpha layer
    shadow = image[3].gaussblur(radius)
    # saturate it but not 100% so there is some anti aliasing
    shadow = shadow * 64
    # recolor shadow
    shadow = shadow.new_from_image(color) \
        .bandjoin(shadow) \
        .copy(interpretation="srgb")
    # composite
    text = shadow.composite(image, "over")
    return text
```
![image](https://user-images.githubusercontent.com/35177464/197420641-ae7941f7-d400-4525-afd1-54271007bdd1.png)

the `shadow * 64` line can be tuned with higher or lower values to sharpen or blur the edge. I still think a dedicated stroke effect or a dilate effect with anti-aliasing would be better but libvips doesn't have those currently

**Reply by jcupitt** [reply]
Ah nice. Yes, I was thinking about this too.

You can remove the `*64` by building the scale into the mask, eg.:

```python
#!/usr/bin/python3

import sys
import pyvips

text = pyvips.Image.text(sys.argv[2], dpi=600)

radius = 5

# we need to make the text image large enough to hold the extra edges
text = text.embed(radius, radius,
    text.width + 2 * radius, text.height + 2 * radius)

# dilate the text with a squared-off gaussian mask
mask = pyvips.Image.gaussmat(radius / 2, 0.1, separable=True)
mask *= 64
wide_text = text.convsep(mask).cast("uchar")

# paint wide text blue
background_text = wide_text \
    .new_from_image([128, 128, 255]) \
    .bandjoin(wide_text) \
    .copy(interpretation="srgb")

# paint original text black
foreground_text = text \
    .new_from_image([0, 0, 0]) \
    .bandjoin(text) \
    .copy(interpretation="srgb")

# composite black over blue
image = background_text.composite(foreground_text, "over")

image.write_to_file(sys.argv[1])
```

**Reply by machineonamission** [reply]
How would I make the mask circular instead of a square?

**Reply by jcupitt** [reply]
Maybe:

```python
#!/usr/bin/python3

import sys
import pyvips

image = pyvips.Image.black(21, 21)
point = pyvips.Image.black(1, 1) + 255
image = image.insert(pyvips.Image.black(1, 1) + 255, 10, 10)
image = image.cast("uchar")

radius = 5

# dilate with a squared-off gaussian mask
mask = pyvips.Image.gaussmat(radius / 2, 0.1, separable=True) * 10
image = image.convsep(mask)

# insert a black pixel to show the start point, and blow up pixels
image = image.insert(pyvips.Image.black(1, 1), 10, 10)
image = image.zoom(5, 5)

image.write_to_file(sys.argv[1])
```

To see the point spread function. I see:

```
$ ./outline2.py x.png
```

![x](https://user-images.githubusercontent.com/580843/197574811-af4a5c07-f4c2-405b-a329-aac75cd75bdd.png)
